### ISSUE TITLE
suggestion for `blocks_in_conditions` generates invalid rust code

### ISSUE THREAD

**Issue by KNSN92** [rust-lang/rust-clippy#17068]
### Summary

`blocks_in_conditions` generates invalid Rust code.
Specifically, the inner match statement is not wrapped in a block, but Clippy tries to generate the `let statement` included code as if it were in a block.
```rust
/** Some(()) => **/ let res = {
    nop();
    42
}; /** match res {
    42 => nop(),
    ...
**/
```

### Lint Name

blocks_in_conditions

### Reproducer

I tried this code:

```rust
fn main() {
    match Some(()) {
        Some(()) => match {
            nop();
            42
        } {
            42 => nop(),
            _ => nop(),
        },
        None => nop(),
    }
}

fn nop() {}
```

I saw this happen:

```
error: expected expression, found `let` statement
 --> src/main.rs:3:21
  |
3 |         Some(()) => let res = {
  |                     ^^^
  |
  = note: only supported directly in conditions of `if` and `while` expressions

error: expected one of `,`, `.`, `?`, `}`, or an operator, found `;`
 --> src/main.rs:6:10
  |
3 |         Some(()) => let res = {
  |                  -- while parsing the `match` arm starting here
...
6 |         }; match res {
  |          ^ expected one of `,`, `.`, `?`, `}`, or an operator

error: aborting due to 2 previous errors

Original diagnostics will follow.

warning: in a `match` scrutinee, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a `let`
 --> src/main.rs:3:21
  |
3 |           Some(()) => match {
  |  _____________________^
4 | |             nop();
5 | |             42
6 | |         } {
  | |_________^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.92.0/index.html#blocks_in_conditions
  = note: `#[warn(clippy::blocks_in_conditions)]` on by default
help: try
  |
3 ~         Some(()) => let res = {
4 +             nop();
5 +             42
6 ~         }; match res {
  |

```

I expected to see this happen:

### Version

```text
rustc 1.95.0 (59807616e 2026-04-14)
binary: rustc
commit-hash: 59807616e1fa2540724bfbac14d7976d7e4a3860
commit-date: 2026-04-14
host: aarch64-apple-darwin
release: 1.95.0
LLVM version: 22.1.2

clippy 0.1.95 (59807616e1 2026-04-14)
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by lanfeust69** [comment]
Here is another example :
```
fn test(b1: bool, b2: bool, i: i32) -> bool {
    assert!(!b1 || i > 42);
    b2 && i <= 42
}

fn main() {
    let (cond1, cond2, cond3) = (false, true, true);
    if cond1 {
        println!("first case");
    } else if {
        let check = test(cond1, cond2, 12);
        eprintln!("check is {check}");
        check && cond3
    } {
        println!("second case");
    } else if cond3 {
        println!("third case");
    } else {
        println!("fourth case");
    }
}
```
where the suggestion doesn't compile.

Moreover, in this case, while I agree that this may not be the nicest code, the only way I can see to avoid the block, when a condition may only be successfully evaluated if we know that the previous one(s) were false, is to stuff all the 2nd, 3rd and 4th cases in a big `else` of the first case, arguably worse.
